cpp/src/arrow/compute/kernels/scalar_hash.cc

```diff
@@ -0,0 +1,221 @@
+      auto extension_type = checked_cast<const ExtensionType*>(array.type);
+      auto storage_array = array;
+      storage_array.type = extension_type->storage_type().get();
+      return HashArray(storage_array, hash_ctx, memory_pool, out);
+    }
+
+    if (type_id == Type::STRUCT) {
+      std::vector<std::shared_ptr<ArrayData>> child_hashes(array.child_data.size());
+      columns.resize(array.child_data.size());
+      for (size_t i = 0; i < array.child_data.size(); i++) {
+        auto child = array.child_data[i];
+        if (is_nested(child.type->id())) {
+          ARROW_ASSIGN_OR_RAISE(child_hashes[i],
+                                HashChild(array, child, hash_ctx, memory_pool));
+          ARROW_ASSIGN_OR_RAISE(columns[i], ToColumnArray(*child_hashes[i], hash_ctx));
+        } else {
+          ARROW_ASSIGN_OR_RAISE(columns[i], ToColumnArray(child, hash_ctx));
+        }
+      }
+      Hasher::HashMultiColumn(columns, hash_ctx, out);
+    } else if (is_list_like(type_id)) {
+      auto values = array.child_data[0];
+      ARROW_ASSIGN_OR_RAISE(auto value_hashes,
+                            HashChild(array, values, hash_ctx, memory_pool));
+      ARROW_ASSIGN_OR_RAISE(
+          columns[0], ToColumnArray(array, hash_ctx, value_hashes->buffers[1]->data()));
+      Hasher::HashMultiColumn(columns, hash_ctx, out);
+    } else {
+      ARROW_ASSIGN_OR_RAISE(columns[0], ToColumnArray(array, hash_ctx));
+      Hasher::HashMultiColumn(columns, hash_ctx, out);
+    }
+    return Status::OK();
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& input_arg, ExecResult* out) {
+    if (input_arg.num_values() != 1 || !input_arg[0].is_array()) {
+      return Status::Invalid("FastHash currently supports a single array input");
+    }
+    ArraySpan hash_input = input_arg[0].array;
+
+    auto exec_ctx = default_exec_context();
+    if (ctx && ctx->exec_context()) {
+      exec_ctx = ctx->exec_context();
+    }
+
+    // Initialize stack-based memory allocator used by Hashing32 and Hashing64
+    util::TempVectorStack stack_memallocator;
+    ARROW_RETURN_NOT_OK(
+        stack_memallocator.Init(exec_ctx->memory_pool(),
+                                3 * sizeof(int32_t) * util::MiniBatch::kMiniBatchLength));
+
+    // Prepare context used by Hashing32 and Hashing64
+    LightContext hash_ctx;
+    hash_ctx.hardware_flags = exec_ctx->cpu_info()->hardware_flags();
+    hash_ctx.stack = &stack_memallocator;
+
+    // Call the hashing function, overloaded based on OutputCType
+    ArraySpan* result_span = out->array_span_mutable();
+    c_type* result_ptr = result_span->GetValues<c_type>(1);
+    ARROW_RETURN_NOT_OK(
+        HashArray(hash_input, &hash_ctx, exec_ctx->memory_pool(), result_ptr));
+
+    return Status::OK();
+  }
+};
+
+const FunctionDoc hash32_doc{
+    "Construct a hash for every element of the input argument",
+    ("An element-wise function that uses an xxHash-like algorithm.\n"
```

Comment:
I don't think we should say anything about being "xxHash-like".

Reply:
Removed.